### PR TITLE
controller/certificate: move cert generation to cert controller

### PR DESCRIPTION
**Pull request by ironcladlou**
Relocate default wildcard certificate generation to the certificate controller to more
clearly decouple certificate management from ingress reconcilication.

Change the certificate controller event handling so that reconciliation is
triggered by clusteringress events. This makes the reconciliation code in the
controller easier to reason about and means we only do cert management if
there's a clusteringress around. It also means we can easily react to
clusteringresses being added/deleted.

Depends on https://github.com/openshift/cluster-ingress-operator/pull/140
Supports https://github.com/openshift/cluster-ingress-operator/pull/139

### CODE REVIEW

**Comment by ironcladlou** [comment]
This one's ready for review now

File: pkg/operator/controller/certificate/ca.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+func (r *reconciler) ensureRouterCASecret() (*corev1.Secret, error) {
+	current, err := r.currentRouterCASecret()
+	if err != nil {
+		return nil, err
+	}
+	if current != nil {
+		return current, nil
+	}
+	desired, err := desiredRouterCASecret(r.operatorNamespace)
```
Comment (pravisankar) [inline pkg/operator/controller/certificate/ca.go]: `desiredRouterCASecret` will generate new cert every time and current != desired after first iteration.
May be we could just use get + create resource pattern that we use in other places to avoid confusion? or we could rename `desired` to something more accurate.

File: pkg/operator/controller/certificate/controller.go [hunk 2]
```diff
@@ -77,119 +59,56 @@ type reconciler struct {
...
-			Namespace: name.Namespace,
-		},
-		Data: map[string][]byte{
-			"tls.crt": certBytes,
-			"tls.key": keyBytes,
-		},
-		Type: corev1.SecretTypeTLS,
 	}
-	return secret, nil
+	return reconcile.Result{}, utilerrors.NewAggregate(errs)
```
Comment (pravisankar) [inline pkg/operator/controller/certificate/controller.go]: return result, ... (otherwise result.RequeueAfter is not propagated to the caller)

File: pkg/operator/controller/certificate/controller.go [hunk 3]
```diff
@@ -77,119 +59,56 @@ type reconciler struct {
...
 }
 
-// createRouterCASecret creates the router CA secret.
-func (r *reconciler) createRouterCASecret(secret *corev1.Secret) error {
-	if err := r.client.Create(context.TODO(), secret); err != nil {
-		return err
+func routerDeploymentName(ci *ingressv1alpha1.ClusterIngress) types.NamespacedName {
```
Comment (pravisankar) [inline pkg/operator/controller/certificate/controller.go]: I think you suggested moving [Resource]Name to controller/names.go or something on slack. Do we want to do that as a separate pr or is it dropped for some other reason?

File: pkg/operator/controller/certificate/ca.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+	if err != nil {
+		return nil, err
+	}
+	if err := r.createRouterCASecret(desired); err != nil {
+		return nil, fmt.Errorf("failed to create CA secret: %v", err)
+	}
+	r.recorder.Event(desired, "Normal", "CreatedWildcardCACert", "Created a default wildcard CA certificate")
```
Comment (Miciah) [inline pkg/operator/controller/certificate/ca.go]: Seems odd to log in `createRouterCASecret` and emit the event in `ensureRouterCASecret`; do we need to do both, and if so, should we do them in the same place?

File: pkg/operator/controller/certificate/controller.go [hunk 5]
```diff
@@ -77,119 +59,56 @@ type reconciler struct {
...
-		if errors.IsNotFound(err) {
-			return nil, nil
+		return reconcile.Result{}, fmt.Errorf("failed to ensure router CA secret: %v", err)
+	}
+
+	result := reconcile.Result{}
+	errs := []error{}
+	ingress := &ingressv1alpha1.ClusterIngress{}
+	if err := r.client.Get(context.TODO(), request.NamespacedName, ingress); err != nil {
+		if !errors.IsNotFound(err) {
```
Comment (Miciah) [inline pkg/operator/controller/certificate/controller.go]: Looks like `!errors.IsNotFound(err)` should be `errors.IsNotFound(err)`.

File: pkg/operator/controller/certificate/default_cert.go [hunk 6]
```diff
@@ -43,12 +43,16 @@ func (r *reconciler) ensureDefaultCertificateForIngress(caSecret *corev1.Secret,
...
 			return fmt.Errorf("failed to delete default certificate: %v", err)
+		} else if deleted {
+			r.recorder.Eventf(ci, "Normal", "DeletedDefaultCertificate", "Deleted default wildcard certificate %q", desired.Name)
 		}
 	case desired != nil && current == nil:
-		if err := r.createRouterDefaultCertificate(desired); err != nil {
+		if created, err := r.createRouterDefaultCertificate(desired); err != nil {
 			return fmt.Errorf("failed to create default certificate: %v", err)
+		} else if created {
+			r.recorder.Eventf(ci, "Normal", "CreatedDefaultCertificate", "Created default wildcard certificate %q", desired.Name)
```
Comment (pravisankar) [inline pkg/operator/controller/certificate/default_cert.go]: What's the guideline for using events over logging?

File: pkg/operator/controller/certificate/ca.go (hunk 4 above)
Comment (ironcladlou) [reply pkg/operator/controller/certificate/ca.go]: I missed this refactor with the other event handling; fixed

File: pkg/operator/controller/certificate/controller.go (hunk 5 above)
Comment (ironcladlou) [reply pkg/operator/controller/certificate/controller.go]: Good catch, fixed

File: pkg/operator/controller/certificate/controller.go (hunk 3 above)
Comment (ironcladlou) [reply pkg/operator/controller/certificate/controller.go]: Started the cleanup of name handling in https://github.com/openshift/cluster-ingress-operator/pull/142; too much work to do here because of import cycles

File: pkg/operator/controller/certificate/default_cert.go (hunk 6 above)
Comment (ironcladlou) [reply pkg/operator/controller/certificate/default_cert.go]: Up to us to decide what's most user friendly AFAIK

File: pkg/operator/controller/certificate/controller.go (hunk 2 above)
Comment (ironcladlou) [reply pkg/operator/controller/certificate/controller.go]: Good catch, fixed

File: pkg/operator/controller/certificate/ca.go (hunk 1 above)
Comment (ironcladlou) [reply pkg/operator/controller/certificate/ca.go]: We're not currently diffing them — are you saying there's an error here, or just that it's confusing?

If desiredRouterCASecret isn't deterministic that seems like a problem, and if it _can't be_ deterministic then our diffing logic would have to account for it (if/when we start diffing)

Comment (Miciah) [reply pkg/operator/controller/certificate/ca.go]: `routerCAConfigMapsEqual` compares only `.data`, similar to how `deploymentConfigChanged` compares only select fields.  I think this pattern is reasonable since there will always be fields that should be ignored (e.g., uid).

Comment (pravisankar) [reply pkg/operator/controller/certificate/ca.go]: There is no error here. It was confusing for me when i saw current can be different from desired (issue, expiry times will differ as those are derived from current time). Ignore if you think this does not confuse others.

Comment (ironcladlou) [reply pkg/operator/controller/certificate/ca.go]: Agree it's weird, and we can fix it in a followup as it's no different than today — I just transferred the logic over as close to the same as possible for the restructuring

**Comment by pravisankar** [comment]
@ironcladlou squash your commits?

File: pkg/operator/controller/certificate/controller.go [hunk 7]
```diff
@@ -77,119 +59,56 @@ type reconciler struct {
...
-	if err := r.client.Get(context.TODO(), name, secret); err != nil {
+	result := reconcile.Result{}
+	errs := []error{}
+	ingress := &ingressv1alpha1.ClusterIngress{}
+	if err := r.client.Get(context.TODO(), request.NamespacedName, ingress); err != nil {
 		if errors.IsNotFound(err) {
-			return nil, nil
+			// This means the ingress was already deleted/finalized and there are
+			// stale queue entries (or something edge triggering from a related
+			// resource that got deleted async).
```
Comment (Miciah) [inline pkg/operator/controller/certificate/controller.go]: > (or something edge triggering from a related resource that got deleted async).

This comment does not make sense in this new context at present time.
